### PR TITLE
Implement uninstaller script

### LINKED ISSUE #239: Make an OS X uninstaller app.
**Branch** Develop

**Target** 0.16.x

**Current Behaviour** We don't have an uninstaller for the OS X app.

**Expected Behaviour** There must be an uninstaller to cleanly remove the OS X app.

Currently we use the `Reset App` button on the `Advanced` tab which cleans and removes .plist files and removes the environment variables.

We need an uninstaller app or a script that can be run by the user.

### PR DESCRIPTION
**Pull request by arceduardvincent**
Hi @aronasorman and @cpauya.
This fixes #239.

**Description:** 
In order that the user can uninstall `KA-Lite.app` and it's independencies we create  `KA-Lite_Uninstall.tool` that can uninstall and remove it's independencies.
I Include the `/Applications/KA-Lite` which bundled `KA-Lite.app` and  `KA-Lite_Uninstall.tool`
Like `Adobe Photoshop` for example they have uninstaller script bundled with the application. In this setup the user can simply locate the uninstaller script and uninstall the `KA-Lite.app`.

**Uninstall Feature Checklist :**
- [x] Include the `KA-Lite_Uninstall.tool` to the `/Applications/KA-Lite` directory .
- [x] Remove `KA-Lite.app` from `/Applications/KA-Lite`.
- [x] Remove .`KA-Lite.app` installed in specific location. 
- [x] Remove the .plist files.
- [x] Confirm if to remove the `KALITE_HOME ` path also (or user wants to preserve the data directory).
- [x] Remove `/usr/bin/kalite` executable.
- [x] Remove `/usr/local/bin/kalite` executable.
- [x] Remove `KALITE_PYTHON` environment variable.
- [x] Remove `/Users/Shared/ka-lite` directory.
- [x] Remove `KALITE_HOME` environment variable. 

**Steps to Replicate:**
 1. Open terminal
 2. Navigate to your `./installer/osx` folder.
 3. Run `./build.sh` to create the `KA-Lite.pkg`
 4. Open the build output `./installers/osx/temp/output/KA-Lite.pkg`
 5. Install the newly built installer.
 6. Once it is installed double click the `/Applications/KA-Lite/KA-Lite_Uninstall.tool` to uninstall the `KA-Lite` and its dependencies.

### CODE REVIEW

**Comment by arceduardvincent** [comment]
Don't merge it yet. Still work on progress

**Comment by djallado** [comment]
@arceduardvincent Done testing this.

After uninstall `KA-Lite`:
 1. `org.learningequality.kalite.plist` was removed.
 2. `/Users/Shared/kalite` was removed
 3. `/Application/KA-Lite` folder was removed
 4. `kalite` executable was removed

Uninstall `KA-Lite` is now working.

**Comment by arceduardvincent** [comment]
Hi @cpauya . I think this feature is ready to go now.

@djallado thanks for the conformation.

**Comment by cpauya** [comment]
@arceduardvincent Please remove `KALITE_HOME` env var too coz the .app also uses it.

File: osx/KA-Lite_Uninstall.tool [hunk 1]
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env bash
+
+# Override any funny stuff from the user.
+export PATH="/bin:/usr/bin:/sbin:/usr/sbin:$PATH"
+
+#----------------------------------------------------------------------
+# Global Variables
+#----------------------------------------------------------------------
+KALITE_MONITOR_APP="/Applications/KA-Lite-Monitor.app"
+KALITE_APP="/Applications/KA-Lite/KA-Lite.app"
```
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: @arceduardvincent Am I correct that we have a `/Applications/KA-Lite/` folder because we now include an uninstall tool?

File: osx/KA-Lite_Uninstall.tool [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+KALITE="kalite"
+KALITE_PLIST="org.learningequality.kalite.plist"
+HOME_LAUNCH_AGENTS="$HOME/Library/LaunchAgents"
+ROOT_LAUNCH_AGENTS="/Library/LaunchAgents"
+LIBRARY_PLIST="$LAUNCH_AGENTS/$KALITE_PLIST"
```
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: `LAUNCH_AGENTS` is not defined anywhere else?

File: osx/KA-Lite_Uninstall.tool [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+#----------------------------------------------------------------------
+# Delete specific location where the kalite is installed.
+# Reliable way for a bash script to get the full path to itself?
+pushd `dirname $0` > /dev/null
+SCRIPTPATH=`pwd`
+popd > /dev/null
+
+# Collect the directories and files to remove
+REMOVE_FILES_ARRAY=()
+append REMOVE_FILES_ARRAY $SCRIPTPATH
```
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: @arceduardvincent Why will you delete files in the current directory where the script is run?

Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: Let's just hard-code the `/Applications/KA-Lite/` for now and use that instead of `pwd` for `SCRIPTPATH`.

**Comment by cpauya** [comment]
@djallado Let's test this today, try running the `KA-Lite_Uninstall.tool` somewhere else (with some unimportant files on it) to see if it behaves like how @EdDixon experienced it today:

> Ed Dixon 8:34 AM
> LOL just a note any content in the folder that script is run from will be DELETED! Regretfully this was my desktop in this case ROFL but it needed a good cleaning anyway ROFL that did seem to work though I am going to reboot and make sure :) brb

I ran it on my `~/Documents/work/installers/osx/` folder and it deleted all files on that folder. :(

File: osx/KA-Lite_Uninstall.tool [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+if [ "$user_input" != "y" ]; then
+    echo "Aborting install. (answer: ${user_input})"
+    key_exit 2
+fi
+
+echo "Do you want to remove the .kalite directory (y/n)?"
+read user_input2
+if [ "$user_input2" == "y" ]; then
+    append REMOVE_FILES_ARRAY "~/.kalite/"
+    echo "Removing .kalite directory (answer: ${user_input2})"
```
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: I suggest you check for `KALITE_HOME` env var if it exists and use that instead.  If not, default it to `~/.kalite/`.

Then use that to prompt the user.

File: osx/KA-Lite_Uninstall.tool [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+fi
+
+echo "The following files and directories will be removed:"
+for file in "${REMOVE_FILES_ARRAY[@]}"; do
+    echo "    $file"
+done
+
```
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: @arceduardvincent Can you add a confirmation prompt here?  This will prevent accidental deletions for the user.

Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: This is a MUST!

File: osx/KA-Lite_Uninstall.tool [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    echo "The uninstallation process requires administrative privileges"
+    echo "because some of the installed files cannot be removed by a"
+    echo "normal user. You will be prompted for a password..."
+    echo ""
+
+    # Use AppleScript so we can use a graphical `sudo` prompt.
+    # This way, people can enter the username they wish to use
+    # for sudo, and it is more Apple-like.
+
+    osascript -e "do shell script \"/bin/rm -Rf ${REMOVE_FILES_ARRAY[*]}\" with administrator privileges"
```
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: Can we have an exclusion checker or list?  We need to make sure some items like "/", "/Library/LaunchAgents/", etc are not deleted by accident!

File: osx/pre_installation.sh [hunk 7]
```diff
@@ -15,9 +15,11 @@
...
 # Global Variables
 #----------------------------------------------------------------------
+KALITE_MONITOR="/Applications/KA-Lite-Monitor.app"
 KALITE="kalite"
 KALITE_PLIST="org.learningequality.kalite.plist"
-LAUNCH_AGENTS="$HOME/Library/LaunchAgents"
+HOME_LAUNCH_AGENTS="$HOME/Library/LaunchAgents"
+ROOT_LAUNCH_AGENTS="/Library/LaunchAgents"
 LIBRARY_PLIST="$LAUNCH_AGENTS/$KALITE_PLIST"
 KALITE_EXECUTABLE_PATH="$(which $KALITE)"
```
Comment (cpauya) [inline osx/pre_installation.sh]: What happens if `kalite` is not found?

**Comment by cpauya** [comment]
Why do we maintain two files (`post_installation.sh` and `KA-Lite_Uninstall.tool`) with similar codes?

File: osx/KA-Lite_Uninstall.tool (hunk 3 above)
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: Or rather, check that the .app itself exists on this location before you include it on the list.

File: osx/KA-Lite_Uninstall.tool [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+function append() {
+    eval $1[\${#$1[*]}]=$2
+}
+
+
+function remove_files_initiator {
+   
+    if which kalite > /dev/null 2>&1; then
+        append REMOVE_FILES_ARRAY $KALITE_EXECUTABLE_PATH
+    fi
```
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: Why is this here?  Why not in the `check_kalite_exc_collector()` function?

File: osx/KA-Lite_Uninstall.tool [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+REMOVE_FILES_ARRAY=()
+append REMOVE_FILES_ARRAY $SCRIPTPATH
+append REMOVE_FILES_ARRAY $HOME_LAUNCH_AGENTS/$KALITE_PLIST
+append REMOVE_FILES_ARRAY $ROOT_LAUNCH_AGENTS/$KALITE_PLIST
+append REMOVE_FILES_ARRAY $KALITE_RESOURCES
+append REMOVE_FILES_ARRAY $KALITE_MONITOR_APP
+append REMOVE_FILES_ARRAY $KALITE_APP
+append REMOVE_FILES_ARRAY $KALITE_USR_BIN_PATH/$KALITE
+append REMOVE_FILES_ARRAY $KALITE_USR_LOCAL_BIN_PATH/$KALITE
+
```
Comment (cpauya) [inline osx/KA-Lite_Uninstall.tool]: Don't assume that those files or directories exists!  Always check they actually exist on the target computer before adding them to the array.

File: osx/pre_installation.sh [hunk 10]
```diff
@@ -92,6 +87,12 @@ function check_kalite_exc_collector {
 ENV=$(env)
 syslog -s -l error "Packages pre-installation initialize with env:'\n'$ENV" 
 
+# Collect the directories and files to remove
+append REMOVE_FILES_ARRAY $HOME_LAUNCH_AGENTS/$KALITE_PLIST
+append REMOVE_FILES_ARRAY $ROOT_LAUNCH_AGENTS/$KALITE_PLIST
+append REMOVE_FILES_ARRAY $KALITE_RESOURCES
+append REMOVE_FILES_ARRAY $KALITE_MONITOR
+
```
Comment (cpauya) [inline osx/pre_installation.sh]: The same as `KA-Lite_Uninstall.tool` - don't assume that those files or directories exists! Always check they actually exist on the target computer before adding them to the array.

**Comment by cpauya** [comment]
@arceduardvincent Please update the documentation regarding this.

**Comment by djallado** [comment]
I ran the script in `~/Desktop/Test/`, with some other file inside the `Test` directory. All the files were deleted inside the `Test` directory.

File: osx/KA-Lite_Uninstall.tool (hunk 5 above)
Comment (arceduardvincent) [inline osx/KA-Lite_Uninstall.tool]: Ok, I will show the files to be remove.

**Comment by djallado** [comment]
@arceduardvincent, I tested your PR and try to copy the `KA-Lite_Uninstall.tool` in `Desktop` with some files and double click that script. The `Uninstall Feature Checklist` above was performed. My files in Desktop was not removed during uninstall.

**Comment by cpauya** [comment]
Thanks, will merge this now.
